### PR TITLE
Feature/sync social get update delete

### PR DESCRIPTION
**Pull request by zong09**
- GET
- PUT
- DELETE

### CODE REVIEW

**Review by julapong-castcle — APPROVED** [review]

File: apps/users/src/app/app.controller.ts [hunk 1]
```diff
@@ -800,6 +824,71 @@ export class UserController {
...
+   */
+  @UsePipes(new ValidationPipe({ skipMissingProperties: true }))
+  @ApiResponse({
+    status: HttpStatus.NO_CONTENT
+  })
+  @ApiBody({
+    type: SocialSyncDeleteDto
+  })
+  @CastleClearCacheAuth(CacheKeyName.SyncSocial)
+  @Delete('syncSocial')
```
Comment (kittipong-castcle) [inline apps/users/src/app/app.controller.ts:874]: By default, `@Delete(...)` returns HTTP status 200 (OK). We should override the return status by adding `@HttpCode(HttpStatus.NO_CONTENT)`

```suggestion
  @Delete('syncSocial')
  @HttpCode(HttpStatus.NO_CONTENT)
```

File: apps/users/src/app/app.controller.ts [hunk 2]
```diff
@@ -800,6 +824,71 @@ export class UserController {
...
+  async deleteSyncSocial(
+    @Req() req: CredentialRequest,
+    @Body() body: SocialSyncDeleteDto
+  ) {
+    logger.log(`Start delete sync social.`);
+    logger.log(JSON.stringify(body));
+    const user = await this._getUserFromIdOrCastcleId(body.castcleId, req);
+    if (user) {
+      await this.socialSyncService.delete(body, user);
+      return '';
```
Comment (kittipong-castcle) [inline apps/users/src/app/app.controller.ts:884]: If this API returns no content (204: NO_CONTENT), the `return` is unnecessary.

```suggestion
```

File: apps/users/src/app/app.controller.ts [hunk 3]
```diff
@@ -800,6 +824,71 @@ export class UserController {
...
+    logger.log(JSON.stringify(body));
+    const user = await this._getUserFromIdOrCastcleId(body.castcleId, req);
+    if (user) {
+      await this.socialSyncService.update(body, user);
+      return '';
+    } else
+      throw new CastcleException(
+        CastcleStatus.FORBIDDEN_REQUEST,
+        req.$language
+      );
```
Comment (kittipong-castcle) [inline apps/users/src/app/app.controller.ts:857]: I think we should return/throw early by writing conditions in form of a guard clause. It's cleaner and easier to read.

```suggestion
    const user = await this._getUserFromIdOrCastcleId(body.castcleId, req);
    
    if (!user) throw new CastcleException(CastcleStatus.FORBIDDEN_REQUEST);
    
    await this.socialSyncService.update(body, user);
```

File: apps/users/src/app/app.controller.ts [hunk 4]
```diff
@@ -800,6 +824,71 @@ export class UserController {
...
+    logger.log(JSON.stringify(body));
+    const user = await this._getUserFromIdOrCastcleId(body.castcleId, req);
+    if (user) {
+      await this.socialSyncService.delete(body, user);
+      return '';
+    } else
+      throw new CastcleException(
+        CastcleStatus.FORBIDDEN_REQUEST,
+        req.$language
+      );
```
Comment (kittipong-castcle) [inline apps/users/src/app/app.controller.ts:889]:
```suggestion
      throw new CastcleException(CastcleStatus.FORBIDDEN_REQUEST);
```

We already implemented the Exception Filter to handle the localization of error messages, so we don't need to pass language to the exception.

Is this correct? @PopCastcle

File: libs/database/src/lib/services/social-sync.service.ts [hunk 5]
```diff
@@ -119,4 +115,64 @@ export class SocialSyncService {
...
+      if (updateSocialSync.uid) socialSync.socialId = updateSocialSync.uid;
+      if (updateSocialSync.userName)
+        socialSync.userName = updateSocialSync.userName;
+      if (updateSocialSync.displayName)
+        socialSync.displayName = updateSocialSync.displayName;
+      if (updateSocialSync.avatar) socialSync.avatar = updateSocialSync.avatar;
+      socialSync.active = updateSocialSync.active;
+      return socialSync.save();
+    } else {
+      this.logger.warn('Cnn not found social sync');
```
Comment (kittipong-castcle) [inline libs/database/src/lib/services/social-sync.service.ts:148]: `Cnn` or `Can`?

**Review by kittipong-castcle — CHANGES_REQUESTED** [review]

**Review by kittipong-castcle — APPROVED** [review]

**Pull request merged**